File: 2020/day_8/second_part.py
from typing import List, NamedTuple, Set


class ProgramState(NamedTuple):
    accumulator: int
    line: int
        

class Instruction:

    def __init__(self, operation: str, argument: int):
        self._operation = operation
        self._argument = argument

    @property
    def operation(self) -> str:
        return self._operation

    @property
    def argument(self) -> int:
        return self._argument

    def __repr__(self) -> str:
        return f"(Instruction: operation='{self._operation}', argument={self.argument})"

    def nop(self, state: ProgramState) -> ProgramState:
        return ProgramState(
            accumulator=state.accumulator, 
            line=state.line+1
        )

    def acc(self, state: ProgramState) -> ProgramState:
        return ProgramState(
            accumulator=state.accumulator + self.argument, 
            line=state.line+1
        )

    def jmp(self, state: ProgramState) -> ProgramState:
        return ProgramState(
            accumulator=state.accumulator, 
            line=state.line + self.argument
        )


class Instructions(List[Instruction]):
    pass


class Program:

    def __init__(self, instructions: Instructions):
        self._instructions = instructions

    @property
    def instructions(self) -> Instructions:
        return self._instructions


    def __iter__(self):
        state = ProgramState(accumulator=0, line=0)
        yield state

        while state.line < len(self._instructions):
            state = self._execute_instruction(self._instructions[state.line], state)
            yield state


    def _execute_instruction(self, instruction: Instruction, state: ProgramState) -> ProgramState:
        return getattr(instruction, instruction.operation)(state)
    

def parse_instruction(entry: str) -> Instruction:
    splitted_entry = entry.split()
    return Instruction(operation=splitted_entry[0], argument=int(splitted_entry[1]))


def get_accumulator_after_end(program: Program) -> int:
    visited_lines: Set[int] = set()
    for state in program:
        if state.line in visited_lines:
            raise Exception('infinite loop')
        else:
            visited_lines.add(state.line)
    return state.accumulator


def find_and_get_accumulator_after_end(program: Program) -> int:
    swaps = {
        'nop': 'jmp',
        'jmp': 'nop'
    }

    instructions = program.instructions
    for line in range(len(instructions)):
        instruction = instructions[line]
        if instruction.operation in swaps:
            new_instructions = Instructions(instructions)
            new_instructions[line] = Instruction(operation=swaps[instruction.operation], argument=instruction.argument)
            try:
                return get_accumulator_after_end(Program(instructions=new_instructions))
            except:
                continue
    raise Exception('no program found')        
        

instructions = Instructions()
entry = input()
while entry != 'end':
    instructions.append(parse_instruction(entry))
    entry = input()

program = Program(instructions)

print("solution:", find_and_get_accumulator_after_end(program))

File: 2020/day_8/haskell/second_part.hs
module Main where

import Data.Array
import qualified Data.Either as Either
import qualified Data.IntSet as IntSet
import System.IO

type Instruction = (String, Int)

type Program = Array Int Instruction

main :: IO ()
main = do
  m <- doReadFile
  case findFirst
    (\p -> Either.isRight $ runProgram p 0 0 IntSet.empty)
    (allPossiblePrograms (parseProgram m) 0) of
    Nothing -> error "No program found"
    Just p -> print $ runProgram p 0 0 IntSet.empty
  return ()

allPossiblePrograms :: Program -> Int -> [Program]
allPossiblePrograms p i
  | i > snd (bounds p) = []
  | op == "jmp" || op == "nop" = [p, newP] ++ allPossiblePrograms p (i + 1)
  | otherwise = allPossiblePrograms p (i + 1)
  where
    (op, val) = p ! i
    newP = p // [(n, (changeOp op, val)) | n <- [i]]

runProgram :: Program -> Int -> Int -> IntSet.IntSet -> Either (Int, Int) Int
runProgram p i acc visited
  | isFinished = Right acc
  | IntSet.member i visited = Left (i, acc)
  | otherwise = runProgram p newI newAcc newVisited
  where
    isFinished = i > snd (bounds p)
    newVisited = IntSet.union (IntSet.singleton i) visited
    (newI, newAcc) = evalInstruction i acc (p ! i)

changeOp :: String -> String
changeOp op
  | op == "jmp" = "nop"
  | op == "nop" = "jmp"
  | otherwise = error $ "Wrong operation: " ++ show op

evalInstruction :: Int -> Int -> Instruction -> (Int, Int)
evalInstruction i acc instruction
  | op == "nop" = (i + 1, acc)
  | op == "jmp" = (i + val, acc)
  | op == "acc" = (i + 1, acc + val)
  | otherwise = error $ "Unknown operator: " ++ show op
  where
    (op, val) = instruction

parseProgram :: [String] -> Program
parseProgram is =
  listArray (0, length instructions - 1) instructions
  where
    instructions = parseInstructions is

parseInstructions :: [String] -> [Instruction]
parseInstructions is
  | null is = []
  | otherwise = parseInstruction (head is) : parseInstructions (tail is)

parseInstruction :: String -> Instruction
parseInstruction i = (op, val)
  where
    (op, s) = break (== ' ') i
    sigval = tail s
    val = m * read (tail sigval)
    m =
      if head sigval == '+'
        then 1
        else -1

findFirst :: (a -> Bool) -> [a] -> Maybe a
findFirst f l
  | null l = Nothing
  | f (head l) = Just (head l)
  | otherwise = findFirst f (tail l)

doReadFile :: IO [[Char]]
doReadFile = withFile "input" ReadMode doReadLine

doReadLine :: Handle -> IO [[Char]]
doReadLine hFile = do
  isEnd <- hIsEOF hFile
  if isEnd
    then return []
    else do
      line <- hGetLine hFile
      rest <- doReadLine hFile
      return (line : rest)
